### ISSUE TITLE
Doesn't detect E string on bass

### ISSUE THREAD

**Issue by JacobJT** [peterkhayes/pitchfinder#24]
It seems that this stops at 40hz, it detects the F1 note on Bass just fine.

I'll be trying to dive in and see if I can figure out where this minimum threshold is coming from, but figured I'd post the issue to see if it's been spotted / you can give some advice on where to look.

Thanks for the tool!

**Comment by JacobJT** [comment]
I suppose this is because bass is tuned an octave lower than guitar? 
https://en.wikipedia.org/wiki/Bass_guitar_tuning

So I assume the math was set up for guitar specifically. I want to use it for guitar and bass, so I'll probably implement a switch to select the instrument.

**Comment by peterkhayes** [comment]
I didn't specifically set any thresholds for guitar or any other instrument. However, it's totally possible that the algorithms break down at low enough frequencies. Can you verify that the error occurs for all of the different algorithms?

**Comment by JacobJT** [comment]
I actually couldn't get the ones other than YIN to work. I forked https://github.com/qiuxiang/react-native-tuner, and was playing around with modifying it.

What I did discover was removing line 46 `bufferSize /= 2;` from yin.js did the trick, and doesn't seem to have broken anything.  I noticed that you were dividing again when setting yinBufferLength again below, and then bufferLength wasn't used again. E string is detected just fine now.

**Comment by peterkhayes** [comment]
If you want to help, one thing you could do would be to add some more example data from real-world sources (as opposed to digital tones) that I could add to the test suite?

**Comment by JacobJT** [comment]
I.e. You want recorded samples of an instrument?

**Comment by peterkhayes** [comment]
Yeah, something that reproduces your problem

**Comment by JacobJT** [comment]
Best I can do at the moment is send you a sound recording of me plucking that E string using my phone, though that is what I was using to pick up the sound in the first place with that tuner app, so I'd imagine it should hit the test case.

Do you happen to have the YIN paper? I followed the link at the top of the file and the PDF link for the YIN paper was a dead link.

**Comment by peterkhayes** [comment]
I don't, sorry :/ I wrote this module awhile ago and haven't done any significant work on it in some time.

**Comment by marksyzm** [comment]
Yup, same on mine, usually even around 150hz I have issues. I can do a push request at some point if this fix mentioned works and run it against the tests if you like?

**Comment by marksyzm** [comment]
81Hz is best I can get from it on an iPhone via vocals but that may be down to how vocal tones are interpreted by the device

**Comment by marksyzm** [comment]
Seems to be an issue with Yin type. I'm using `pitchy` these days which uses the McLeod pitch method: 
https://github.com/ianprime0509/pitchy
https://github.com/sevagh/pitch-detection/blob/master/misc/mcleod/README.md

I imagine this should be fairly straightforward to translate into this library?

**Comment by peterkhayes** [comment]
The Macleod method already exists I think - does it have issues?

https://github.com/peterkhayes/pitchfinder/blob/master/src/detectors/macleod.ts

**Comment by marksyzm** [comment]
Ah! I assumed it wasn't there as the readme didn't say so! I'd recommend this for the lower frequency issues people have mentioned

**Comment by peterkhayes** [comment]
Oh good call that it's not included in the readme! I think someone else added it awhile back but probably only did the code?

**Comment by marksyzm** [comment]
I can confirm that it works more accurately and faster than pitchy. AMDF does work, also, taking around 4ms to calculate per buffer. Definitely worth putting in the Readme.

**Comment by peterkhayes** [comment]
Updated!

**Comment by marksyzm** [comment]
Sorry, just one more thing as I'm testing this recently but the McLeod method is more accurate down to 32hz as long as you use 1024 length buffer - 512 can just about do it. Frequency range is affected by buffer length, as can be expected when you consider how a waveform works. It is also well worth borrowing the work done in pitchy.js for FFT optimising with McLeod, which is around 4 times faster:

https://github.com/ianprime0509/pitchy
